File: backend/routes/ghostpass.py
from fastapi import APIRouter, Depends, HTTPException
from supabase import Client
from database import get_db
from routes.auth import get_current_user
from models import PurchaseRequest, PurchaseResponse, GhostPass
from typing import List
import logging

router = APIRouter(prefix="/ghostpass", tags=["GhostPass"])
logger = logging.getLogger(__name__)

# Pricing in cents: 1 Day ($10), 3 Days ($20), 7 Days ($50)
PRICES = {
    1: 1000,   # $10.00
    3: 2000,   # $20.00  
    7: 5000    # $50.00
}

@router.post("/purchase", response_model=PurchaseResponse)
def purchase_pass(
    req: PurchaseRequest,
    user=Depends(get_current_user),
    db: Client = Depends(get_db)
):
    """Purchase a GhostPass with specified duration"""
    if req.duration not in PRICES:
        raise HTTPException(status_code=400, detail="Invalid duration. Must be 1, 3, or 7 days")
    
    price_cents = PRICES[req.duration]
    
    try:
        logger.info(f"Attempting to purchase pass for user {user.id}, duration: {req.duration} days, price: {price_cents} cents")
        
        # Ensure user exists in our users table (should already be done by auth, but double-check)
        db.table("users").upsert({
            "id": str(user.id),
            "email": user.email
        }, on_conflict="id").execute()
        
        # Check if user has a wallet and sufficient balance
        wallet_response = db.table("wallets").select("*").eq("user_id", str(user.id)).execute()
        if not wallet_response.data:
            raise HTTPException(status_code=404, detail="No wallet found. Please fund your wallet first.")
        
        wallet = wallet_response.data[0]
        logger.info(f"Wallet found: balance_cents={wallet['balance_cents']}, required={price_cents}")
        if wallet["balance_cents"] < price_cents:
            raise HTTPException(
                status_code=400, 
                detail=f"Insufficient balance. Required: ${price_cents/100:.2f}, Available: ${wallet['balance_cents']/100:.2f}"
            )
        
        # Atomic purchase via database function
        # This checks balance, deducts amount, creates pass, and logs transaction
        pass_id = None
        try:
            logger.info("Making RPC call to purchase_pass")
            
            # Try the RPC call with error handling
            rpc_params = {
                "p_user_id": str(user.id),
                "p_amount": price_cents,
                "p_duration_days": req.duration
            }
            logger.info(f"RPC params: {rpc_params}")
            
            result = db.rpc("purchase_pass", rpc_params).execute()
            
            # The function now returns a table with one row containing pass_id
            if hasattr(result, 'data') and result.data and len(result.data) > 0:
                # Extract pass_id from first row of result
                pass_id = str(result.data[0]['pass_id'])
                logger.info(f"RPC call successful - Pass created with ID: {pass_id}")
            else:
                raise Exception("No pass ID returned from purchase_pass function")
            
        except Exception as e:
            error_str = str(e)
            logger.error(f"RPC call failed: {error_str}")
            logger.error(f"Trying manual approach as fallback")
            pass_id = None
            
            # Manual transaction approach as fallback (only if pass_id wasn't extracted)
            if pass_id is None:
                logger.error(f"Trying manual approach as fallback")
                try:
                    from datetime import datetime, timedelta
                    import uuid
                    
                    # Generate pass ID
                    pass_id = str(uuid.uuid4())
                    expires_at = datetime.utcnow() + timedelta(days=req.duration)
                    
                    # Deduct from wallet
                    wallet_update = db.table("wallets")\
                        .update({"balance_cents": wallet["balance_cents"] - price_cents})\
                        .eq("id", wallet["id"])\
                        .execute()
                    
                    if not wallet_update.data:
                        raise Exception("Failed to update wallet balance")
                    
                    # Create ghost pass
                    pass_insert = db.table("ghost_passes")\
                        .insert({
                            "id": pass_id,
                            "user_id": str(user.id),
                            "status": "ACTIVE",
                            "expires_at": expires_at.isoformat()
                        })\
                        .execute()
                    
                    if not pass_insert.data:
                        raise Exception("Failed to create ghost pass")
                    
                    # Log transaction with balance tracking
                    current_balance = wallet["balance_cents"]
                    new_balance = current_balance - price_cents
                    
                    transaction_insert = db.table("transactions")\
                        .insert({
                            "wallet_id": wallet["id"],
                            "type": "SPEND",
                            "amount_cents": -price_cents,
                            "balance_before_cents": current_balance,
                            "balance_after_cents": new_balance,
                            "vendor_name": "GhostPass System",
                            "metadata": {
                                "pass_id": pass_id,
                                "duration_days": req.duration,
                                "expires_at": expires_at.isoformat()
                            }
                        })\
                        .execute()
                    
                    logger.info(f"Manual transaction completed - Pass ID: {pass_id}")
                    
                except Exception as manual_error:
                    logger.error(f"Manual transaction also failed: {manual_error}")
                    raise HTTPException(status_code=500, detail="Purchase failed. Please try again.")

        
        # Fetch the created pass to get expiration
        try:
            logger.info(f"Fetching pass details for ID: {pass_id}")
            pass_response = db.table("ghost_passes")\
                .select("expires_at")\
                .eq("id", str(pass_id))\
                .single()\
                .execute()
            
            logger.info(f"Pass response type: {type(pass_response)}")
            logger.info(f"Pass response: {pass_response}")
            
            # Check if pass_response has data attribute
            if hasattr(pass_response, 'data'):
                logger.info(f"Pass response.data type: {type(pass_response.data)}")
                logger.info(f"Pass response.data: {pass_response.data}")
                
                if pass_response.data and isinstance(pass_response.data, dict):
                    expires_at = pass_response.data.get("expires_at")
                    if expires_at:
                        logger.info(f"Found expires_at: {expires_at}")
                    else:
                        raise Exception("No expiration date found in pass data")
                else:
                    raise Exception(f"Pass data is not a dict: {type(pass_response.data)}")
            else:
                raise Exception("Pass response has no data attribute")
            
        except Exception as e:
            logger.error(f"Error fetching pass details: {e}")
            logger.error(f"Error type: {type(e)}")
            # Fallback: calculate expiration manually
            from datetime import datetime, timedelta
            expires_at = (datetime.utcnow() + timedelta(days=req.duration)).isoformat()
            logger.info(f"Using fallback expires_at: {expires_at}")
        
        return PurchaseResponse(
            status="success",
            pass_id=pass_id,
            expires_at=expires_at,
            amount_charged_cents=price_cents
        )
        
    except HTTPException:
        # Re-raise HTTPExceptions as-is (don't convert to 500)
        raise
    except Exception as e:
        error_msg = str(e)
        logger.error(f"Pass purchase error: {error_msg}")
        logger.error(f"Error type: {type(e)}")
        
        # Handle specific database errors
        if "Insufficient balance" in error_msg:
            raise HTTPException(status_code=400, detail="Insufficient wallet balance")
        elif "Wallet not found" in error_msg:
            raise HTTPException(status_code=404, detail="Wallet not found. Please fund your wallet first.")
        elif "function purchase_pass" in error_msg.lower() and "does not exist" in error_msg.lower():
            raise HTTPException(status_code=500, detail="Database function not found. Please run database setup.")
        else:
            raise HTTPException(status_code=500, detail=f"Purchase failed: {error_msg}")

@router.get("/status", response_model=GhostPass)
def get_pass_status(
    user=Depends(get_current_user),
    db: Client = Depends(get_db)
):
    """Get user's current active pass status"""
    try:
        logger.info(f"Fetching pass status for user: {user.id}")
        
        # Get the most recent active pass
        pass_response = db.table("ghost_passes")\
            .select("*")\
            .eq("user_id", str(user.id))\
            .eq("status", "ACTIVE")\
            .gte("expires_at", "now()")\
            .order("created_at", desc=True)\
            .limit(1)\
            .execute()
        
        logger.info(f"Pass query result: {pass_response.data}")
        
        if not pass_response.data:
            logger.info("No active pass found")
            raise HTTPException(status_code=404, detail="No active pass found")
        
        active_pass = pass_response.data[0]
        logger.info(f"Returning active pass: ID={active_pass.get('id')}, expires_at={active_pass.get('expires_at')}")
        
        return active_pass
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Pass status error: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch pass status")

@router.get("/passes", response_model=List[GhostPass])
def get_user_passes(
    user=Depends(get_current_user),
    db: Client = Depends(get_db),
    active_only: bool = False
):
    """Get user's GhostPasses"""
    try:
        query = db.table("ghost_passes").select("*").eq("user_id", str(user.id))
        
        if active_only:
            query = query.eq("status", "ACTIVE").gte("expires_at", "now()")
        
        passes_response = query.order("created_at", desc=True).execute()
        return passes_response.data
        
    except Exception as e:
        logger.error(f"Passes fetch error: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch passes")

@router.get("/passes/{pass_id}", response_model=GhostPass)
def get_pass_details(
    pass_id: str,
    user=Depends(get_current_user),
    db: Client = Depends(get_db)
):
    """Get specific pass details"""
    try:
        pass_response = db.table("ghost_passes")\
            .select("*")\
            .eq("id", pass_id)\
            .eq("user_id", str(user.id))\
            .single()\
            .execute()
        
        if not pass_response.data:
            raise HTTPException(status_code=404, detail="Pass not found")
        
        return pass_response.data
        
    except Exception as e:
        logger.error(f"Pass details error: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch pass details")

File: ghost-pass/routes/ghostpass.ts
import express from 'express';
import type { Request, Response } from 'express';
import { getDb } from '../database.ts';
import { getCurrentUser } from '../auth.ts';
import type { PurchaseRequest, PurchaseResponse } from '../types.ts';
import { v4 as uuidv4 } from 'uuid';

const router = express.Router();

// Default pricing in cents
const DEFAULT_PRICES: Record<number, number> = {
  1: 1000,   // $10.00
  3: 2000,   // $20.00
  5: 3500,   // $35.00
  7: 5000,   // $50.00
  10: 6500,  // $65.00
  14: 8500,  // $85.00
  30: 10000  // $100.00
};

const getPricing = async (db: any) => {
  try {
    const config = await db.from('system_configs').select('config_value').eq('config_key', 'ghostpass_pricing');
    if (config.data && config.data[0]?.config_value) {
      const pricing = config.data[0].config_value;
      return Object.fromEntries(
        Object.entries(pricing).map(([k, v]) => [parseInt(k), parseInt(v as string)])
      );
    }
  } catch (error) {
    console.warn('Failed to fetch pricing config:', error);
  }
  return DEFAULT_PRICES;
};

// Get pricing (public endpoint)
router.get('/pricing', async (req: Request, res: Response) => {
  try {
    const db = getDb();
    const pricing = await getPricing(db);
    res.json({
      pricing,
      currency: 'USD'
    });
  } catch (error) {
    console.error('Pricing fetch error:', error);
    res.status(500).json({ detail: 'Failed to fetch pricing' });
  }
});

// Purchase ghost pass
router.post('/purchase', getCurrentUser, async (req: Request, res: Response) => {
  const user = (req as any).user;
  const { duration }: PurchaseRequest = req.body;

  try {
    const db = getDb();
    const PRICES = await getPricing(db);

    if (!PRICES[duration]) {
      return res.status(400).json({ detail: `Invalid duration. Must be one of: ${Object.keys(PRICES).join(', ')}` });
    }

    const price_cents = PRICES[duration];

    console.log(`Attempting to purchase pass for user ${user.id}, duration: ${duration} days, price: ${price_cents} cents`);

    // Ensure user exists
    await db.from('users').upsert({
      id: user.id,
      email: user.email
    }, { onConflict: 'id' });

    // Check wallet
    const walletResponse = await db.from('wallets').select('*').eq('user_id', user.id);
    if (!walletResponse.data || walletResponse.data.length === 0) {
      return res.status(404).json({ detail: 'No wallet found. Please fund your wallet first.' });
    }

    const wallet = walletResponse.data[0];
    if (wallet.balance_cents < price_cents) {
      return res.status(400).json({
        detail: `Insufficient balance. Required: $${(price_cents / 100).toFixed(2)}, Available: $${(wallet.balance_cents / 100).toFixed(2)}`
      });
    }

    // Generate pass ID
    const pass_id = uuidv4();
    const expires_at = new Date(Date.now() + duration * 24 * 60 * 60 * 1000).toISOString();

    // Deduct from wallet
    const newBalance = wallet.balance_cents - price_cents;
    await db.from('wallets').update({
      balance_cents: newBalance
    }).eq('id', wallet.id);

    // Create ghost pass
    const passInsert = await db.from('ghost_passes').insert({
      id: pass_id,
      user_id: user.id,
      status: 'ACTIVE',
      expires_at,
      duration_days: duration,
      price_cents
    }).select();

    if (!passInsert.data) {
      return res.status(500).json({ detail: 'Failed to create ghost pass' });
    }

    // Log transaction
    await db.from('transactions').insert({
      wallet_id: wallet.id,
      type: 'SPEND',
      amount_cents: -price_cents,
      balance_before_cents: wallet.balance_cents,
      balance_after_cents: newBalance,
      vendor_name: 'GhostPass System',
      metadata: {
        pass_id,
        duration_days: duration,
        expires_at
      }
    });

    const response: PurchaseResponse = {
      pass_id,
      expires_at,
      amount_charged_cents: price_cents,
      status: 'success'
    };

    res.json(response);
  } catch (error) {
    console.error('Purchase error:', error);
    res.status(500).json({ detail: 'Purchase failed' });
  }
});

// Get pass status
router.get('/status', getCurrentUser, async (req: Request, res: Response) => {
  const user = (req as any).user;

  try {
    const db = getDb();
    const passResponse = await db
      .from('ghost_passes')
      .select('*')
      .eq('user_id', user.id)
      .eq('status', 'ACTIVE')
      .gte('expires_at', new Date().toISOString())
      .order('created_at', { ascending: false })
      .limit(1);

    if (!passResponse.data || passResponse.data.length === 0) {
      return res.status(404).json({ detail: 'No active pass found' });
    }

    res.json(passResponse.data[0]);
  } catch (error) {
    console.error('Pass status error:', error);
    res.status(500).json({ detail: 'Failed to fetch pass status' });
  }
});

// Get all passes for user
router.get('/passes', getCurrentUser, async (req: Request, res: Response) => {
  const user = (req as any).user;
  const { active_only } = req.query;

  try {
    const db = getDb();
    let query = db.from('ghost_passes').select('*').eq('user_id', user.id);

    if (active_only === 'true') {
      query = query.eq('status', 'ACTIVE').gte('expires_at', new Date().toISOString());
    }

    const passesResponse = await query.order('created_at', { ascending: false });

    res.json(passesResponse.data || []);
  } catch (error) {
    console.error('Passes fetch error:', error);
    res.status(500).json({ detail: 'Failed to fetch passes' });
  }
});

// Get specific pass details
router.get('/passes/:pass_id', getCurrentUser, async (req: Request, res: Response) => {
  const user = (req as any).user;
  const { pass_id } = req.params;

  try {
    const db = getDb();
    const passResponse = await db
      .from('ghost_passes')
      .select('*')
      .eq('id', pass_id)
      .eq('user_id', user.id)
      .single();

    if (!passResponse.data) {
      return res.status(404).json({ detail: 'Pass not found' });
    }

    res.json(passResponse.data);
  } catch (error) {
    console.error('Pass details error:', error);
    res.status(500).json({ detail: 'Failed to fetch pass details' });
  }
});

// Ghost Pass Modes routes
router.get('/contexts', (req: Request, res: Response) => {
  res.json({
    contexts: {
      'default': {
        name: 'Default Context',
        modes: ['standard'],
        pricing: { 1: 1000, 3: 2000, 7: 5000 },
        description: 'Standard GhostPass access'
      }
    },
    timestamp: new Date().toISOString()
  });
});

router.post('/check-context', (req: Request, res: Response) => {
  const { context_name } = req.body;
  res.json({
    context_name,
    valid: true,
    modes_available: ['standard'],
    message: 'Context is valid'
  });
});

router.post('/interact', getCurrentUser, (req: Request, res: Response) => {
  res.json({
    success: true,
    interaction_type: 'access',
    message: 'Interaction successful'
  });
});

export default router;
